Userland/Utilities/pls.cpp

```diff
@@ -0,0 +1,228 @@
+        warnln("Failed to unveil paths!");
+        return 1;
+    }
+
+    // Lock veil
+    unveil(nullptr, nullptr);
+
+    // Check the permissions and owner of /etc/plsusers. This ensures the integrity of the file.
+    struct stat sudoers_stat;
+    if (stat("/etc/plsusers", &sudoers_stat) < 0) {
+        perror("stat");
+        return 1;
+    }
+
+    if (sudoers_stat.st_mode != EXPECTED_PERMS) {
+        warnln("/etc/plsusers permissions are incorrect! (has the file been compromised?!)");
+        return 4;
+    }
+
+    if (sudoers_stat.st_uid != 0 && sudoers_stat.st_gid != 0) {
+        warnln("/etc/plsusers is not owned by root! (has the file been compromised?!)");
+        return 4;
+    }
+
+    const char* username = getlogin();
+    auto sudoers_file_or_error = Core::File::open("/etc/plsusers", Core::IODevice::ReadOnly);
+    bool user_found = false;
+    if (sudoers_file_or_error.is_error()) {
+        warnln("couldn't open /etc/plsusers! ({})", sudoers_file_or_error.error());
+        return 1;
+    }
+
+    for (auto line = sudoers_file_or_error.value()->line_begin(); !line.at_end(); ++line) {
+        auto line_str = *line;
+
+        // Skip any comments
+        if (line_str.starts_with("#"))
+            continue;
+
+        // Our user is in the plsusers file!
+        if (line_str.to_string() == username) {
+            user_found = true;
+            break;
+        }
+    }
+
+    // User isn't in the plsusers file
+    if (!user_found) {
+        warnln("{} is not in the plsusers file!", username);
+        return 2;
+    }
+
+    // The user was in the plsusers file, now let's ask for their password to ensure that it's actually them...
+    uid_t current_uid = getuid();
+    auto account_or_error = Core::Account::from_name(username);
+
+    if (account_or_error.is_error()) {
+        warnln("Core::Account::from_name: {}", account_or_error.error());
+        return 1;
+    }
+
+    const auto& account = account_or_error.value();
+    if (current_uid != 0 && account.has_password()) {
+        auto password = Core::get_password();
+        if (password.is_error()) {
+            warnln("{}", password.error());
+            return 1;
+        }
+
+        if (!account.authenticate(password.value().characters())) {
+            warnln("Incorrect or disabled password.");
+            return 1;
+        }
+    }
+
+    // TODO: Support swapping users instead of just defaulting to root
+    if (setgid(0) < 0) {
+        perror("setgid");
+        return 1;
+    }
+
+    if (setuid(0) < 0) {
+        perror("setuid");
+        return 1;
+    }
+
+    // Build the arguments list passed to `execvpe`
+    Vector<const char*> exec_args;
+    for (const auto& arg : command) {
+        exec_args.append(arg);
+    }
+
+    // Always terminate with a NULL (to signal end of args list)
+    exec_args.append(nullptr);
+
+    // Build the environment arguments
+    StringBuilder builder;
+
+    // Build SUDO_USER envvar
```

Comment:
> Also why is it necessary?

There are instances where a program invoked by `pls` and running as `root` may wish to know the original user ID (usually for logging/auditing, instead of logging all actions as `root`).

`sudo` does something similar:

```
$ whoami && sudo sh -c 'echo $SUDO_USER'
user
user
```

As an example, consider a `sudo` configuration which permits users to run only a limited set of commands, one of which is a program which requires elevated privileges and stores information related to the user. This program has no method to determine the user name/uid/gid/etc of the calling user, only the name/uid/gid/etc of `root`.

Currently, unlike `sudo`, `pls` does not support the same level of control over which commands can be executed by authorised users - it allows either all or nothing. Presumably eventually the `plsusers` configuration will be updated to allow this level of granularity.

Related:

* https://unix.stackexchange.com/questions/190221/sudo-user-undefined-in-sudo-manual-clarification-would-be-lovely
* https://stackoverflow.com/questions/10272784/how-do-i-get-the-users-real-uid-if-the-program-is-run-with-sudo

Reply:
I'm removing these for now until we have something a bit more solid.